### DISCUSSION TITLE
Drift database access hanging specifically when being served from S3

### DISCUSSION THREAD

**Discussion by justro** [simolus3/drift#2563] [General]
I am running into an issue that seems to be specifically occurring when I am serving a flutter app from S3. I didn't want to file a bug yet since I don't have a minimal repro and am unsure if this is a drift problem or (maybe more likely) has to do with how I have my S3 deployment configured. I will be super grateful for any advice/suggestions anyone has.

I have a flutter app that I build for iOS, Android, and web. It uses drift as a local store. I have:

```
environment:
  sdk: ">=2.19.0 <3.0.0"
  flutter: 3.3.10
.......
drift: ^2.10.0
sqlite3: ^2.1.0
sqlite3_flutter_libs: ^0.5.15
......
(dev deps)
drift_dev: ^2.10.0
```

My web folder has drift_worker.dart 2.10 and sqlite3.wasm 2.1 pulled from the repos linked in the web instructions. (I currently am using the sqlite3.debug.wasm build actually, but I don't know how to get any information from it, not sure if I need to set some flag to get logs or what?)

When I load my app, it initializes the DB, and eventually will perform some requests against it. On native platforms, this works without issue. 

On web, if I run locally it works, whether I build in debug or release mode.

When I run on web, specifically when it's being served from an aws cloudfront/s3 deployment, it seems to hang on initialization.  Running from a local web build, built release or otherwise, db initialization seems to proceed as expected. The hang happens whichever browser I've tried (from mac chrome/safari or windows edge/firefox), and whether I'm in normal or private mode.

For example, using [the code here](https://drift.simonbinder.eu/web/#migrating-from-custom-wasmdatabases), on a working (local) run, I can see log lines dropped from the initializeDatabase function, but when I'm loading from a cloudfront/s3 distribution, I do not see the log lines happen.

Basically, any async call, including the database open, seem to never return when served from cf/s3, but work fine from others.

I had seen similar behavior when using older versions of drift, but it seemed to happen really infrequently, and when I used an incognito window or cleared local data, it worked fine. Having upgraded to 2.10 recently, I run into the error 100% of the time when deployed from s3.

It sounds like maybe it's a header configuration issue. I'm not seeing the CORS headers I expect coming from cf/s3, so I will check into that next.

**Comment by justro** [comment]
Sorry, another note, FWIW whether it hangs or not, the browsers I'm using seem to fallback to sharedIndexedDB: 

`│ 💡 Using WasmStorageImplementation.sharedIndexedDb due to missing browser features: {MissingBrowserFeature.dedicatedWorkersInSharedWorkers, MissingBrowserFeature.sharedArrayBuffers}`

Happens whether it works or not, so maybe not relevant, but thought perhaps the specific storage impl might be worth knowing.

**Comment by justro** [comment]
After further troubleshooting, it seems like I was wrong, it *does* depend on my build environment. The broken build is done in AWS codepipeline, but if I copy my local sources into the s3 bucket for the deployment, problem seems to be resolved.

**Comment by simolus3** [comment]
> (I currently am using the sqlite3.debug.wasm build actually, but I don't know how to get any information from it

It should print them to the console.

> `Using WasmStorageImplementation.sharedIndexedDb`

If you're not enabling CORS headers, this is actually expected.

Do you see logs about workers not being loaded or something similar that would point towards a possible cause?

**Reply by justro** [reply]
Thanks so much for responding!

I do not see any errors that would indicate the service workers are failing. Nothing in console seems to indicate an error. Looking at the Application tab in devtools shows the flutter SW as green. In firefox, I was also able to see the shared drift workers, and at least it's not obvious to me if there's a problem.

* I copied the failing bits out of the s3 buckets to a local folder. Ran the same code locally, it works. No longer think it's a build env issue, but apparently something about how it's served?

* I added access-control-allow-origin: *,  cross-origin-opener-policy: same-origin, cross-origin-embedder-policy: require-corp. I can see that the selected implementation switches over to opfs, but the same behavior persists.

I have cloned drift, and added a bunch of print statements to the drift_worker.js, just to see if I could pin down something there. So far, it only is just showing me that local and remote seem to be following more or less the same process. I'll paste some logs here...this is logging coming from my own app, and my compiled drift_worker with the print statements.

Here is the working (running local) version

```
┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
js_primitives.dart:42 │ #0   Error
js_primitives.dart:42 │ #1       at Object.lk (http://localhost:49860/main.dart.js:8385:24)
js_primitives.dart:42 ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
js_primitives.dart:42 │ 💡 Opening database
js_primitives.dart:42 └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Service Worker was updated because "Update on reload" was checked in the DevTools Application panel.
drift_worker.js:4959 calling workerMainForOpen
drift_worker.js:4959 workerMainForOpen: done
drift_worker.js:4959 WasmDatabase.workerMainForOpen: DedicatedDriftWorker(self).start()
drift_worker.js:4959 dedicated worker started
drift_worker.js:4959 workerMainForOpen: done
drift_worker.js:4959 workerMainForOpen returned
drift_worker.js:4959 Received message from worker: [object Object]
drift_worker.js:4959 RequestCompatibilityCheck.fromJsPayload goal_setting
drift_worker.js:4959 dedicated worker Received message: Instance of 'minified:dp'
drift_worker.js:4959 Requesting compatibility check for goal_setting
drift_worker.js:4959 compatibility result:   supportsNestedWorkers: true,  canAccessOpfs: true,  supportsIndexedDb: true,  supportsSharedArrayBuffers: false,  opfsExists: false,  indexedDbExists: false,  existingDatabases: []
js_primitives.dart:42 ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
js_primitives.dart:42 │ #0   Error
js_primitives.dart:42 │ #1       at Object.lk (http://localhost:49860/main.dart.js:8385:24)
js_primitives.dart:42 ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
js_primitives.dart:42 │ 💡 Database opened
js_primitives.dart:42 └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
js_primitives.dart:42 ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
js_primitives.dart:42 │ #0   Error
js_primitives.dart:42 │ #1       at Object.lk (http://localhost:49860/main.dart.js:8385:24)
js_primitives.dart:42 ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
js_primitives.dart:42 │ 💡 Using WasmStorageImplementation.sharedIndexedDb due to missing browser features: {MissingBrowserFeature.dedicatedWorkersInSharedWorkers, MissingBrowserFeature.sharedArrayBuffers}
js_primitives.dart:42 └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
js_primitives.dart:42 ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
js_primitives.dart:42 │ #0   Error
js_primitives.dart:42 │ #1       at Object.lk (http://localhost:49860/main.dart.js:8385:24)
js_primitives.dart:42 ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
js_primitives.dart:42 │ 💡 Initializing database
js_primitives.dart:42 └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```
the above will continue with my normal application logging, which indicates that the db is responding and we are retrieving and creating values.

And here is the hanging (serving from aws cloudfront+s3) version:

```
┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
main.dart.js:43003 │ #0   Error
main.dart.js:43003 │ #1       at Object.lk (https://app.dev.focused.space/main.dart.js:8385:24)
main.dart.js:43003 ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
main.dart.js:43003 │ 💡 Opening database
main.dart.js:43003 └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Service Worker was updated because "Update on reload" was checked in the DevTools Application panel.
drift_worker.js:4959 calling workerMainForOpen
drift_worker.js:4959 workerMainForOpen: done
drift_worker.js:4959 WasmDatabase.workerMainForOpen: DedicatedDriftWorker(self).start()
drift_worker.js:4959 dedicated worker started
drift_worker.js:4959 workerMainForOpen: done
drift_worker.js:4959 workerMainForOpen returned
drift_worker.js:4959 Received message from worker: [object Object]
drift_worker.js:4959 RequestCompatibilityCheck.fromJsPayload goal_setting
drift_worker.js:4959 dedicated worker Received message: Instance of 'minified:dp'
drift_worker.js:4959 Requesting compatibility check for goal_setting
drift_worker.js:4959 compatibility result:   supportsNestedWorkers: true,  canAccessOpfs: true,  supportsIndexedDb: true,  supportsSharedArrayBuffers: false,  opfsExists: false,  indexedDbExists: false,  existingDatabases: []
main.dart.js:43003 ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
main.dart.js:43003 │ #0   Error
main.dart.js:43003 │ #1       at Object.lk (https://app.dev.focused.space/main.dart.js:8385:24)
main.dart.js:43003 ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
main.dart.js:43003 │ 💡 Database opened
main.dart.js:43003 └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
main.dart.js:43003 ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
main.dart.js:43003 │ #0   Error
main.dart.js:43003 │ #1       at Object.lk (https://app.dev.focused.space/main.dart.js:8385:24)
main.dart.js:43003 ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
main.dart.js:43003 │ 💡 Using WasmStorageImplementation.sharedIndexedDb due to missing browser features: {MissingBrowserFeature.dedicatedWorkersInSharedWorkers, MissingBrowserFeature.sharedArrayBuffers}
main.dart.js:43003 └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```
In this case, the logging stops.

The " Initializing database" line in the working log is coming from the initializeDatabase method passed into WasmDatabase.open...this is more or less copied from drift docs:

```
/// Obtains a database connection for running drift on the web.
DatabaseConnection connect() {
  return DatabaseConnection.delayed(
    Future(() async {
      _logger.i('Opening database');
      final result = await WasmDatabase.open(
        databaseName:
            'goal_setting', // prefer to only use valid identifiers here
        sqlite3Uri: Uri.parse('/sqlite3.debug.wasm'),
        driftWorkerUri: Uri.parse('/drift_worker.js'),
        initializeDatabase: () async {
          _logger.i('Initializing database');
          // Manually open the file system previously used
          final fs = await IndexedDbFileSystem.open(dbName: 'goal_setting');
          const oldPath =
              'goal_setting.db'; // The path passed to WasmDatabase before

          Uint8List? oldDatabase;

          // Check if the old database exists
          if (fs.xAccess(oldPath, 0) != 0) {
            // It does, then copy the old file
            final openedRes = fs.xOpen(Sqlite3Filename(oldPath), 0);
            final file = openedRes.file;
            final blob = Uint8List(file.xFileSize());
            file
              ..xRead(blob, 0)
              ..xClose();
            fs.xDelete(oldPath, 0);

            oldDatabase = blob;
          }

          await fs.close();
          return oldDatabase;
        },
      );
      _logger.i('Database opened');

      if (result.missingFeatures.isNotEmpty) {
        // Depending how central local persistence is to your app, you may want
        // to show a warning to the user if only unrealiable implemetentations
        // are available.
        _logger.i('Using ${result.chosenImplementation} due to missing browser '
            'features: ${result.missingFeatures}');
      }

      return result.resolvedExecutor;
    }),
  );
}
```

So it seems key to me that in the working implementation, this method is being called, but there's no indication that it's being executed in the failing case. I may have to make personal builds of the drift libs to get any more logging about why that's not happening. Hopefully I'm not totally on the wrong track.

I am going to look into that, and possibly also work on trying to reproduce the behavior outside of my app.

Again, very much appreciate any thoughts or suggestions.

**Reply by justro** [reply]
Sorry just another note about the logs, please ignore where the logging says:

```
main.dart.js:43003 │ #0   Error
main.dart.js:43003 │ #1       at Object.lk (https://app.dev.focused.space/main.dart.js:8385:24)
```

I think this is just because the logger package needs src maps or something on web, in any case it shows up in both working and non-working envs.

**Reply by simolus3** [reply]
It's not exactly clear to me where it's going wrong either. I think there _might_ be a deadlock happening because you're re-using the IndexedDB database name? You're using `goal_setting` for the old and the new database.

You could try also adding debug logs to `DriftServerController.serve` in  `drift/lib/src/web/wasm_setup/shared.dart` - in particular before and after the `await switch` to see if it maybe gets stuck there.

**Reply by justro** [reply]
After about 6 days of troubleshooting, I think I finally figured out what's going on here.  By default, when an aws cloudfront deployment serves a wasm file out of s3, it serves it as content-type: application/octet-stream. I added a behavior to the deployment to serve it as application/wasm, and it immediately started working. I did not know about chrome://inspect for getting logs out of shared workers, so I was not seeing the log lines I had in the shared worker...

It turns out, DriftServerController.serve gets stuck at this line:

        final sqlite3 = await WasmSqlite3.loadFromUrl(message.sqlite3WasmUri);

when the content type is not application/wasm. 

I had been using an older version of drift (2.5.0) and associated worker/sqlite3.wasm builds, and hadn't run into this problem with those older, experimental versions. I'm not sure if I would consider this a bug or what, although some kind of error message or something if the content-type is going to be an issue would be great.

This is a really great library and I deeply appreciate the time and effort that must have gone into this...especially since I've gotten a better idea of how much there is to it. Please let me know if you'd like me to file an issue or provide any more details, document it, etc.

Thanks a ton!

**Reply by simolus3** [reply]
Good catch! And I completely agree that this is a bug in drift. The worker should catch the wasm file not loading and properly return this error to the main document. These kinds of mistakes are easy to make and drift should therefore be very explicit about what went wrong (instead of just doing nothing).

You can open an issue for this if you want to track this bug, but I'll also keep it in mind from this discussion and I'll hopefully fix it soon.

**Reply by justro** [reply]
Thanks so much! I won't bother with a bug unless it's helpful to you...if so, lmk, otherwise, thanks again!

**Reply by simolus3** [reply]
After some further investigation, I think this is happening because the error is getting reported asynchronously, and we only try to load the wasm file when the database is first used (e.g. the first statement to it is run).

That's how drift generally works. I've tried moving this to an earlier place but that would require changing the worker protocol which would be pretty painful to do (since drift needs to support older worker versions as well). In my testing, I was able to reliably catch the error by creating an instance of the app database class and calling `await db.executor.ensureOpen(db)`. This will open the database and run all migrations, throwing if that fails. Otherwise, you'd get the error at the first statement, which might not be handled correctly.
